Initial Data Loading and Cleaning

In [11]:
#libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

In [12]:
#define relative path to dataset
path_train = '/Users/sumaiyauddin/Documents/ML/Project/Project Repo/sign_mnist_train.csv'
path_test = '/Users/sumaiyauddin/Documents/ML/Project/Project Repo/sign_mnist_test.csv'

#define image resolution 
res = (28,28)

In [13]:
#saving data into pandas dataframes 
test = pd.read_csv(path_test)
train = pd.read_csv(path_train)

In [14]:
#printing shape
print(f'Train Shape:{train.shape}')
print(f'Test Shape:{test.shape}')


Train Shape:(27455, 785)
Test Shape:(7172, 785)


The training dataset consists of 27455 28x28 pixel grayscale images, including the associated label. The testing dataset consists of 7172 28x28 images. The label and pixels are split into X and y for both training and testing data.

In [15]:
#training data 
X_train = train.iloc[:,1:]
y_train = train.iloc[:,0]

#testing data
X_test = test.iloc[:,1:]
y_test = test.iloc[:,0]

The below is a function to find the indices of a specified label.

In [16]:
#fuction to find the indices given a label 
def find_indices(data,label):
    #check if data is numpy array
    if type(data) == np.ndarray:
        #return indices
        return np.where(data==label)
    #check if data is pandas series 
    elif type(data) == pd.Series:
        #return indices
        return data[data==label].index
    #else not supported in this function
    else:
        raise Exception('Not supported data type for this function.')

Creating a dictionary where key is the numeric label in the dataset and value is the corresponding letter.

In [17]:
#letters
letters = ['A','B','C','D','E','F','G','H','I','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y']
#numbers 
numbers = sorted(list(y_train.unique()))
#dictionary of labels 
labels = dict(zip(numbers,letters))

Inspecting examples for each letter in the dataset.

## Naive Bayes

In [20]:
# Train and predict using Naive Bayes
gnb = GaussianNB()
y_pred = gnb.fit(X_train, y_train).predict(X_test)

In [21]:
# Print accuracy and classification report
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")
print("Classification report:\n", classification_report(y_test, y_pred, target_names=letters))


Accuracy: 0.3898494143892917
Classification report:
               precision    recall  f1-score   support

           A       0.71      0.48      0.57       331
           B       0.96      0.40      0.56       432
           C       0.72      0.50      0.59       310
           D       0.60      0.40      0.48       245
           E       0.53      0.56      0.55       498
           F       0.39      0.26      0.31       247
           G       0.54      0.56      0.55       348
           H       0.85      0.39      0.53       436
           I       0.22      0.45      0.30       288
           K       0.27      0.41      0.33       331
           L       0.41      0.45      0.43       209
           M       0.37      0.19      0.25       394
           N       0.38      0.31      0.34       291
           O       0.37      0.49      0.42       246
           P       0.43      0.81      0.56       347
           Q       0.29      0.60      0.39       164
           R       0.13     

## Logistic regration